helix-core/src/test/java/org/apache/helix/participant/TestDistControllerElection.java

```diff
@@ -197,35 +218,42 @@ public void testCompeteLeadership() throws Exception {
 
     // Create controller leaders
     final Map<String, ZKHelixManager> managerList = new HashMap<>();
+    final List<GenericHelixController> controllers = new ArrayList<>();
     for (int i = 0; i < managerCount; i++) {
       String controllerName = "controller_" + i;
       ZKHelixManager manager =
           new ZKHelixManager(clusterName, controllerName, InstanceType.CONTROLLER, ZK_ADDR);
       GenericHelixController controller0 = new GenericHelixController();
       DistributedLeaderElection election =
           new DistributedLeaderElection(manager, controller0, Collections.EMPTY_LIST);
+      controllers.add(controller0);
       manager.connect();
       managerList.put(manager.getInstanceName(), manager);
     }
 
-    // Remove leader manager one by one, and verify if the leader node exists
-    while(!managerList.isEmpty()) {
-      // Ensure a controller successfully acquired leadership.
-      Assert.assertTrue(TestHelper.verify(new TestHelper.Verifier() {
-        @Override
-        public boolean verify() {
-          LiveInstance liveInstance = accessor.getProperty(keyBuilder.controllerLeader());
-          if (liveInstance != null) {
-            // disconnect the current leader manager
-            managerList.remove(liveInstance.getInstanceName()).disconnect();
-            return true;
-          } else {
-            return false;
+    try {
+      // Remove leader manager one by one, and verify if the leader node exists
+      while (!managerList.isEmpty()) {
+        // Ensure a controller successfully acquired leadership.
+        Assert.assertTrue(TestHelper.verify(new TestHelper.Verifier() {
+          @Override
+          public boolean verify() {
+            LiveInstance liveInstance = accessor.getProperty(keyBuilder.controllerLeader());
+            if (liveInstance != null) {
+              // disconnect the current leader manager
+              managerList.remove(liveInstance.getInstanceName()).disconnect();
+              return true;
+            } else {
+              return false;
+            }
           }
-        }
-      }, 1000));
+        }, 1000));
+      }
+    } finally {
+      for (GenericHelixController controller : controllers) {
+        controller.shutdown();
```

Comment:
How about the managers? Please disconnect them too.

Reply:
good point. added.